lib/internal/mime.js

```diff
@@ -0,0 +1,403 @@
+  return pattern;
+}
+
+const NotHTTPTokenCodePoint = hardenRegExp(/[^!#$%&'*+\-.^_`|~A-Za-z0-9]/g);
+const NotHTTPQuotedStringCodePoint = hardenRegExp(/[^\t\u0020-~\u0080-\u00FF]/g);
+
+const END_BEGINNING_WHITESPACE = hardenRegExp(/[^\r\n\t ]|$/);
+const START_ENDING_WHITESPACE = hardenRegExp(/[\r\n\t ]*$/);
+
+function toASCIILower(str) {
+  let ret = '';
+  for (const c of str) {
+    if (c >= 'A' && c <= 'Z') {
+      ret += StringPrototypeToLowerCase(c);
+    } else {
+      ret += c;
+    }
+  }
+  return ret;
+}
+
+const SOLIDUS = '/';
+const SEMICOLON = ';';
+function parseTypeAndSubtype(str) {
+  // Skip only HTTP whitespace from start
+  let position = SafeStringPrototypeSearch(str, END_BEGINNING_WHITESPACE);
+  // read until '/'
+  const typeEnd = StringPrototypeIndexOf(str, SOLIDUS, position);
+  const trimmedType = typeEnd === -1 ?
+    StringPrototypeSlice(str, position) :
+    StringPrototypeSlice(str, position, typeEnd);
+  const invalidTypeIndex = SafeStringPrototypeSearch(trimmedType,
+                                                     NotHTTPTokenCodePoint);
+  if (trimmedType === '' || invalidTypeIndex !== -1 || typeEnd === -1) {
+    throw new ERR_INVALID_MIME_SYNTAX('type', str, invalidTypeIndex);
+  }
+  // skip type and '/'
+  position = typeEnd + 1;
+  const type = toASCIILower(trimmedType);
+  // read until ';'
+  const subtypeEnd = StringPrototypeIndexOf(str, SEMICOLON, position);
+  const rawSubtype = subtypeEnd === -1 ?
+    StringPrototypeSlice(str, position) :
+    StringPrototypeSlice(str, position, subtypeEnd);
+  position += rawSubtype.length;
+  if (subtypeEnd !== -1) {
+    // skip ';'
+    position += 1;
+  }
+  const trimmedSubtype = StringPrototypeSlice(
+    rawSubtype,
+    0,
+    SafeStringPrototypeSearch(rawSubtype, START_ENDING_WHITESPACE));
+  const invalidSubtypeIndex = SafeStringPrototypeSearch(trimmedSubtype,
+                                                        NotHTTPTokenCodePoint);
+  if (trimmedSubtype === '' ||
+    invalidSubtypeIndex !== -1) {
+    throw new ERR_INVALID_MIME_SYNTAX('subtype', str, trimmedSubtype);
+  }
+  const subtype = toASCIILower(trimmedSubtype);
+  const target = ObjectCreate(null);
+  target.type = type;
+  target.subtype = subtype;
+  target.parametersStringIndex = position;
+  return target;
+}
+
+const EQUALS_SEMICOLON_OR_END = hardenRegExp(/[;=]|$/);
+const QUOTED_VALUE_PATTERN = hardenRegExp(/^(?:([\\]$)|[\\][\s\S]|[^"])*(?:(")|$)/u);
+
+function removeBackslashes(str) {
+  let ret = '';
+  // We stop at str.length - 1 because we want to look ahead one character.
+  let i;
+  for (i = 0; i < str.length - 1; i++) {
+    const c = str[i];
+    if (c === '\\') {
+      i++;
+      ret += str[i];
+    } else {
+      ret += c;
+    }
+  }
+  // We add the last character if we didn't skip to it.
+  if (i === str.length - 1) {
+    ret += str[i];
+  }
+  return ret;
+}
+
+function parseParametersString(str, position, paramsMap = new Map()) {
```

Comment:
```suggestion
function parseParametersString(str, position, paramsMap = new SafeMap()) {
```
if paramsMap ever gets returned to the user, it should be `new Map(paramsMap.entries())` first.

Reply:
I think line 21 of the file means that `Map` === `SafeMap` but maybe we shouldn't do that and keep `SafeMap` separate from `Map` to avoid confusion.

Reply:
oh snap, you're right. but yes, i think it's best to keep the SafeMap name for clarity.